File: advent2017/extras/cxx/src/days/day02.cpp
#include <algorithm>
#include <chrono>
#include <cctype>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <openssl/sha.h>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

const std::string kExpectedSha = "c64165a1af8ab4877e736a095bde2b22d523468077099fab5a338f53b0059681";
const std::string kExpectedPart1 = "36174";
const std::string kExpectedPart2 = "244";

std::string resolve_input(const std::string& provided) {
    if (!provided.empty()) return provided;
    const std::vector<std::string> cands{
        "advent2017/Day2/d2_input.txt",
        "Day2/d2_input.txt",
        "../Day2/d2_input.txt",
        "../../Day2/d2_input.txt",
    };
    for (const auto& c : cands) {
        if (std::filesystem::exists(c)) return c;
    }
    throw std::runtime_error("input not found");
}

std::string read_all(const std::string& path) {
    std::ifstream in(path, std::ios::binary);
    if (!in) throw std::runtime_error("failed to read input");
    return std::string((std::istreambuf_iterator<char>(in)), {});
}

std::string sha256_hex(const std::string& data) {
    unsigned char out[SHA256_DIGEST_LENGTH];
    SHA256(reinterpret_cast<const unsigned char*>(data.data()), data.size(), out);
    std::ostringstream oss;
    oss << std::hex << std::setfill('0');
    for (auto b : out) oss << std::setw(2) << static_cast<int>(b);
    return oss.str();
}

std::vector<std::vector<int>> parse_rows(const std::string& raw) {
    std::vector<std::vector<int>> rows;
    std::istringstream in(raw);
    std::string line;
    while (std::getline(in, line)) {
        bool all_ws = true;
        for (char c : line) {
            if (!std::isspace(static_cast<unsigned char>(c))) {
                all_ws = false;
                break;
            }
        }
        if (all_ws) continue;

        std::istringstream ls(line);
        std::vector<int> row;
        int v = 0;
        while (ls >> v) row.push_back(v);
        if (!row.empty()) rows.push_back(std::move(row));
    }
    return rows;
}

int solve_part1(const std::vector<std::vector<int>>& rows) {
    int total = 0;
    for (const auto& row : rows) {
        auto mm = std::minmax_element(row.begin(), row.end());
        total += (*mm.second - *mm.first);
    }
    return total;
}

int solve_part2(const std::vector<std::vector<int>>& rows) {
    int total = 0;
    for (const auto& row : rows) {
        bool found = false;
        for (std::size_t i = 0; i < row.size() && !found; ++i) {
            for (std::size_t j = 0; j < row.size(); ++j) {
                if (i == j) continue;
                const int a = row[i];
                const int b = row[j];
                if (b != 0 && a % b == 0) {
                    total += a / b;
                    found = true;
                    break;
                }
            }
        }
        if (!found) throw std::runtime_error("no divisible pair found");
    }
    return total;
}

}  // namespace

int main(int argc, char** argv) {
    int part = 0;
    std::string input;
    for (int i = 1; i < argc; ++i) {
        std::string a = argv[i];
        if (a == "--part") part = std::stoi(argv[++i]);
        else if (a == "--input") input = argv[++i];
        else throw std::runtime_error("unknown arg");
    }
    if (part != 1 && part != 2) throw std::runtime_error("--part 1|2 required");

    input = resolve_input(input);
    const std::string raw = read_all(input);
    if (sha256_hex(raw) != kExpectedSha) throw std::runtime_error("checksum mismatch");
    const auto rows = parse_rows(raw);

    auto t0 = std::chrono::steady_clock::now();
    std::string ans;
    std::string expected;
    if (part == 1) {
        ans = std::to_string(solve_part1(rows));
        expected = kExpectedPart1;
    } else {
        ans = std::to_string(solve_part2(rows));
        expected = kExpectedPart2;
    }
    if (ans != expected) throw std::runtime_error("answer mismatch");

    std::cout << ans << "\n";
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::cerr << "[cxx-fancy] day=2 part=" << part << " runtime_ms=" << ms << "\n";
    return 0;
}

File: advent2017/extras/rust/src/bin/day02.rs
use sha2::{Digest, Sha256};
use std::env;
use std::fs;
use std::path::Path;
use std::time::Instant;

const EXPECTED_SHA: &str = "c64165a1af8ab4877e736a095bde2b22d523468077099fab5a338f53b0059681";
const EXPECTED_PART1: &str = "36174";
const EXPECTED_PART2: &str = "244";

fn resolve_input(provided: &str) -> String {
    if !provided.is_empty() {
        return provided.to_string();
    }
    let cands = [
        "advent2017/Day2/d2_input.txt",
        "Day2/d2_input.txt",
        "../Day2/d2_input.txt",
        "../../Day2/d2_input.txt",
    ];
    for c in cands {
        if Path::new(c).exists() {
            return c.to_string();
        }
    }
    panic!("input not found");
}

fn sha256_file(path: &str) -> String {
    let b = fs::read(path).expect("read");
    let mut h = Sha256::new();
    h.update(&b);
    hex::encode(h.finalize())
}

fn parse_rows(raw: &str) -> Vec<Vec<i64>> {
    raw.lines()
        .filter_map(|line| {
            let line = line.trim();
            if line.is_empty() {
                None
            } else {
                Some(
                    line.split_whitespace()
                        .map(|x| x.parse::<i64>().expect("int"))
                        .collect::<Vec<i64>>(),
                )
            }
        })
        .collect::<Vec<Vec<i64>>>()
}

fn solve_part1(rows: &[Vec<i64>]) -> i64 {
    rows.iter()
        .map(|row| row.iter().max().unwrap() - row.iter().min().unwrap())
        .sum()
}

fn solve_part2(rows: &[Vec<i64>]) -> i64 {
    let mut total: i64 = 0;
    for row in rows {
        let mut found = false;
        for (i, a) in row.iter().enumerate() {
            for (j, b) in row.iter().enumerate() {
                if i == j {
                    continue;
                }
                if *b != 0 && a % b == 0 {
                    total += a / b;
                    found = true;
                    break;
                }
            }
            if found {
                break;
            }
        }
        if !found {
            panic!("no divisible pair found");
        }
    }
    total
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let mut part = 0;
    let mut input = String::new();
    let mut i = 1;
    while i < args.len() {
        match args[i].as_str() {
            "--part" => {
                part = args[i + 1].parse().unwrap();
                i += 2;
            }
            "--input" => {
                input = args[i + 1].clone();
                i += 2;
            }
            _ => panic!("unknown arg"),
        }
    }
    if part != 1 && part != 2 {
        panic!("--part 1|2 required")
    }

    let in_path = resolve_input(&input);
    if sha256_file(&in_path) != EXPECTED_SHA {
        panic!("checksum mismatch");
    }
    let raw = fs::read_to_string(&in_path).expect("read_to_string");
    let rows = parse_rows(&raw);

    let t0 = Instant::now();
    let (ans, expected) = if part == 1 {
        (solve_part1(&rows).to_string(), EXPECTED_PART1)
    } else {
        (solve_part2(&rows).to_string(), EXPECTED_PART2)
    };

    if ans != expected {
        panic!("answer mismatch");
    }
    println!("{}", ans);
    let ms = t0.elapsed().as_nanos() as f64 / 1e6;
    eprintln!("[rust-fancy] day=2 part={} runtime_ms={:.3}", part, ms);
}
